experimental/rule_inference/base_prompt.py

```diff
@@ -0,0 +1,131 @@
+    
+    ========================= Output Format =========================
+    
+    <file_name_start> your_rules_file.toml <file_name_end>
+    ```toml
+    # Define your rule within this section
+    [[rules]]
+    # Provide a unique name for your rule
+    name = "your_rule_name"
+    
+    # Write a Tree-sitter query to identify the code pattern for refactoring
+    query = """(
+        (method_invocation name: (_) @name
+                           arguments: (_) @args) @invk
+        (#eq? @name "some_string"))
+    """
+    
+    # Specify the captured node from the query that will be replaced
+    replace_node = "invk"
+    
+    # Replacement string that will substitute `replace_node`
+    replace = "X.other_string(@args)"
+    
+    # Specify any placeholders in your queries that will be filled in at runtime
+    holes = ["hole1", "hole2"]
+    
+    # Specify if this rule should be triggered first. If it depends on other rules, set to false
+    is_seed_rule = true
+    
+    # If necessary, define filters for your rule
+    [[rules.filters]]
+    # This pattern should match any ancestor of the captured node (optional)
+    enclosing_node = "(your_enclosing_node_pattern) @your_capture_name"
+    
+    # Define patterns that should not be present within the enclosing_node (optional)
+    not_contains = [
+        """
+        (your_not_contains_pattern) @your_capture_name
+        """
+    ]
+    # Define a pattern that should be present within the enclosing_node (optional)
+    contains =
+        """
+        (your_contains_pattern) @your_capture_name
+        """
+    # Define the minimum and maximum number of children that should match the 'contains' pattern (optional)
+    at_least = 1
+    at_most = 5
+    ```
+    
+    ========================= Rule Examples =========================
+    '''
+    input_template = """
+    ========================= Task =========================
+
+    ### Original Code ###
+    
+    {source_code}
+    
+    ### Tree-sitter Representation of Original Code ###
+    
+    {source_tree}
+    
+    ### Refactored Code ###
+    
+    {target_code}
+    
+    ### Tree-sitter Representation of Refactored Code ###
+    
+    {target_tree}
+    
+    ========================= Output =========================
+    """
+
+    @staticmethod
+    def generate_prompt(**kwargs):
+        examples = BasePrompt._get_examples("../../src/cleanup_rules")
+        formatted = (
+            BasePrompt.explanation
+            + "\n"
+            + examples
+            + "\n"
+            + BasePrompt.input_template.format(**kwargs)
+        )
+        return [{"role": "user", "content": formatted}]
+
+    @staticmethod
+    def _get_examples(path_to_examples_rules):
+        task_examples = ""
+        for root, dirs, files in os.walk(path_to_examples_rules):
+            for file in files:
+                if file.endswith("rules.toml"):
```

Comment:
Does this imply that currently `PiranhaAgent` does not understand the concept of edges?

Reply:
Yes

Reply:
(for now)